### PR TITLE
Placeholder dependency resolution callback

### PR DESCRIPTION
**Pull request by Rheeseyb**
**Problem:**
During a discussion we realised that our resolving of placeholder dependencies was actually very broken, and would not support importing assets to use directly in the Preview would not work, since we were only supporting side effects when importing placeholder dependencies.

**Fix:**
Support a callback to be used whenever the running code requires a placeholder dependency, allowing us to explicitly force a re-render or re-evaluation after that dependency has been downloaded. This also gives us the option to extend the behaviour to display something to signify to the user that the editor is downloading further resources (I have left 2 `// FIXME` s in the code to mark that this has _not_ been done as part of this work).

I've also refactored some of the uses of this code to support better test coverage.

**Commit Details:**
- `fetchMissingFileDependency` now returns a `Promise` containing the newly downloaded modules
- `getEditorRequireFn` now takes a callback `onRemoteModuleDownload` to allow us to explicitly deal with running code triggering a download, as well as the completion of that download
- Split out the core of `property-controls-info.tsx` into `property-controls-processor.ts` to allow testing, specifically so that I could ensure the above change was correctly implemented in the property controls processing
- Updated the preview to now trigger a re-render when a placeholder dependency has been finished downloading
- Tests (and some other refactoring to support them)!

### CODE REVIEW

File: editor/src/components/editor/npm-dependency/npm-dependency.ts [hunk 1]
```diff
@@ -333,6 +333,20 @@ export function dependenciesWithEditorRequirements(
...
 
+export function orderedDependenciesEqual(
+  l: Array<RequestedNpmDependency>,
+  r: Array<RequestedNpmDependency>,
+): boolean {
+  return (
+    l.length === r.length &&
+    l.reduce(
+      (matches: boolean, next: RequestedNpmDependency, index: number) =>
+        matches && shallowEqual(next, r[index]),
```
Comment (seanparsons) [inline editor/src/components/editor/npm-dependency/npm-dependency.ts:344]: Why `shallowEqual`? Does that run the risk of a change to `RequestedNpmDependency` breaking unexpectedly?

Comment (Rheeseyb) [reply editor/src/components/editor/npm-dependency/npm-dependency.ts:344]: Hmmm tbh we could probably just use `fastDeepEquals` and bin off this function

**Review by seanparsons — APPROVED** [review]

File: editor/src/components/editor/npm-dependency/npm-dependency.ts (hunk 1 above)
Comment (Rheeseyb) [reply editor/src/components/editor/npm-dependency/npm-dependency.ts]: Fixed in a647bdbb

File: editor/src/core/es-modules/package-manager/package-manager.spec.ts [hunk 2]
```diff
@@ -134,13 +122,14 @@ describe('ES Dependency Manager — Real-life packages', () => {
       fail(`Expected successful nodeModules fetch`)
     }
     const nodeModules = fetchNodeModulesResult.nodeModules
-    const req = getRequireFn(moduleUpdater(nodeModules), nodeModules)
+    const onRemoteModuleDownload = jest.fn()
```
Comment (balazsbajorics) [inline editor/src/core/es-modules/package-manager/package-manager.spec.ts:125]: you don't seem to be passing `onRemoteModuleDownload` anywhere, so the `expect(onRemoteModuleDownload).toBeCalledTimes(0)` is not really testing anything

Comment (Rheeseyb) [reply editor/src/core/es-modules/package-manager/package-manager.spec.ts:125]: Ooops

File: editor/src/core/property-controls/property-controls-processor.spec.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import * as antdPackagerResponse from '../es-modules/test-cases/antd-packager-response.json'
+import { PackagerServerResponse, requestedNpmDependency } from '../shared/npm-dependency-types'
+import { initPropertyControlsProcessor } from './property-controls-processor'
+import { fullNodeModulesUpdate, getThirdPartyPropertyControls } from './property-controls-utils'
+
+require('jest-fetch-mock').enableMocks()
+
+const simpleCssContent = '.utopiaClass { background-color: red; }'
+
+beforeEach(() => {
```
Comment (balazsbajorics) [inline editor/src/core/property-controls/property-controls-processor.spec.ts:18]: I wonder if we should make this beforeEach a project-level jest configuration

File: editor/src/core/es-modules/package-manager/package-manager.spec.ts (hunk 2 above)
Comment (Rheeseyb) [reply editor/src/core/es-modules/package-manager/package-manager.spec.ts:125]: Fixed

File: editor/src/core/property-controls/property-controls-processor.spec.ts (hunk 3 above)
Comment (Rheeseyb) [reply editor/src/core/property-controls/property-controls-processor.spec.ts:18]: Yeah I wonder about this too. The question is where do you draw the line here? We don't want every test performing a load of side effects before each run but at the same point we don't want to have to deal with forgetting to run the specific required side effects in the test files that need them to run. I think with something like this it should be lightweight enough to not matter if we run it before every test :shrug:

Comment (balazsbajorics) [reply editor/src/core/property-controls/property-controls-processor.spec.ts:18]: I think `resetDepPackagerCache` passes the lightweightness test, it sets a variable to `{}`

**Review by balazsbajorics — APPROVED** [review]

**Pull request merged**